clients/src/main/java/org/apache/kafka/common/record/DefaultRecordBatch.java

```diff
@@ -500,7 +508,19 @@ public Record next() {
                 throw new NoSuchElementException();
 
             readRecords++;
-            return readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            Record rec = readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
+            if (readRecords == numRecords) {
+                // validate the batch
+                boolean isValid = false;
+                try {
+                    readNext(baseOffset, baseTimestamp, baseSequence, logAppendTime);
```

Comment:
Yeah, @ijuma is right. Unfortunately, we may have to live without this check until we have a better input interface to work with. For a little additional context, we have found that the overhead of throwing `EOFException` is excessive and we took pains to eliminate it, so it would be unfortunate if we still need it in the end.

The consequence of not having this check is that a buggy or malicious producer could write additional data at the end of the batch which would not be reachable. It pollutes the log and may be a difficult issue for a client developer to debug, but it doesn't actually cause any problems that I can think of, so maybe we can live with it.